src/utils.cc

```diff
@@ -106,4 +111,151 @@ void InitDefaultConfig(Config *config) {
+  }
+
+  return result;
+}
+
+u64 UpdateDeviceState(PlacementSchema &schema,
+                      TargetViewState &node_state) {
+  u64 result {0};
+  node_state.ordered_cap.clear();
+
+  for (auto [size, target] : schema) {
+    result += size;
+    node_state.bytes_available[target.bits.device_id] -= size;
+    node_state.ordered_cap.insert(
+      std::pair<u64, TargetID>(
+        node_state.bytes_available[target.bits.device_id], target));
+  }
+
+  return result;
+}
+
+void PrintNodeState(TargetViewState &node_state) {
+  for (int i {0}; i < node_state.num_devices; ++i) {
+    std::cout << "  capacity of device[" << i << "]: "
+              << node_state.bytes_available[i]
+              << '\n' << std::flush;
+    std::cout << "  available ratio of device["<< i << "]: "
+              << static_cast<double>(node_state.bytes_available[i])/
+                 node_state.bytes_capacity[i]
+              << "\n\n" << std::flush;
+  }
+}
+
+std::vector<TargetID> GetDefaultTargets(size_t n) {
+  std::vector<TargetID> result(n);
+  for (size_t i = 0; i < n; ++i) {
+    TargetID id = {};
+    id.bits.node_id = 1;
+    id.bits.device_id = (DeviceID)i;
+    id.bits.index = i;
+    result[i] = id;
+  }
+
+  return result;
+}
+
+std::pair<size_t, size_t> GetBlobBound(BlobSizeRange blob_size_range) {
+  if (blob_size_range == BlobSizeRange::kSmall)
+    return {0, KILOBYTES(64)};
+  else if (blob_size_range == BlobSizeRange::kMedium)
+    return {KILOBYTES(64), MEGABYTES(1)};
+  else if (blob_size_range == BlobSizeRange::kLarge)
+    return {MEGABYTES(1), MEGABYTES(4)};
+  else if (blob_size_range == BlobSizeRange::kXLarge)
+    return {MEGABYTES(4), MEGABYTES(64)};
+  else if (blob_size_range == BlobSizeRange::kHuge)
+    return {GIGABYTES(1), GIGABYTES(1)};
+
+  std::cout << "No specified blob range is found.\n"
+            << "Use small blob range (0, 64KB].";
+  return {0, KILOBYTES(64)};
+}
+
+std::vector<size_t> GenFixedTotalBlobSize(size_t total_size,
+                                          BlobSizeRange range) {
+  std::vector<size_t> result;
+  size_t used_size {};
+  size_t size {};
+  std::random_device dev;
+  std::mt19937 rng(dev());
+
+  std::pair bound = GetBlobBound(range);
+  size_t lo_bound  = bound.first;
+  size_t hi_bound = bound.second;
+
+  while (used_size < total_size) {
+    std::vector<hermes::api::Blob> input_blobs;
+    if (total_size - used_size > hi_bound) {
+      std::uniform_int_distribution<std::mt19937::result_type>
+        distribution(lo_bound, hi_bound);
+      size = distribution(rng);
+      used_size += size;
+    } else {
+      size = total_size - used_size;
+      used_size = total_size;
+    }
+    result.push_back(size);
+  }
+  return result;
+}
+
+std::vector<size_t> GenFixedNumberOfBlobs(int num,
+                                          size_t each_blob_size) {
+  std::vector<size_t> result;
+
+  for (auto i {0}; i < num; ++i) {
+    result.push_back(each_blob_size);
+  }
+  return result;
+}
```

Comment:
This functionality can be achieved more efficiently (in terms of amount of code and number of instructions) with the `std::vector` constructor that takes a size and a default value: `std::vector<size_t> vec(num, each_blob_size);`. I would just replace all calls to `GetFixedNumberOfBlobs(num, each_blob_size);` with `std::vector<T>(num, each_blob_size)`.

Reply:
Right, this function is silly...